### PR TITLE
Ldap new synchronization field

### PR DESCRIPTION
**Pull request by wawax**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #number

Add a new synchronization field for ldap users.
Add unit tests for AuthLDAP & Authldapreplicate

WIP, please do not merge

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -67,3 +74,8 @@ notifications:
 branches:
   except:
     - /^0\..*$/
+
+addons:
```
Comment (BanzaiMan) [inline .travis.yml]: This definition of `addons` nullifies the previous definition of `addons`. You probably mean:

```yaml
addons:
  apt:
    packages:
    - ldap-utils
    - slapd
  hosts:
    - ldap-master
    - ldap-slave
```

Comment (BanzaiMan) [inline .travis.yml]: Furthermore, each jobs in `matrix.include` defines `addons`, which will override these, so this is effectively a noop.

If you want to use these addons, you'd need to add this to each of the jobs in `matrix.include`.

Comment (wawax) [inline .travis.yml]: thank you ! let's try it :)

**Comment by trasher** [comment]
Rely on #2310

**Comment by trasher** [comment]
Require https://github.com/glpi-project/glpi/pull/2604 to be merged.

File: tests/units/AuthLdap.php [hunk 2]
```diff
@@ -850,4 +865,251 @@ public function testLdapImportUserGroup() {
...
+      $this->array($user->fields)
+         ->string['name']->isIdenticalTo('brazil7test')
+         ->string['user_dn']->isIdenticalTo('uid=brazil7test,ou=people,ou=ldap3,dc=glpi,dc=org');
+
+      $this->boolean($auth->user_present)->isTrue();
+      //FIXME?
+      //$this->string($auth->user_dn)->isIdenticalTo($user->fields['user_dn']);
+      $this->resource($auth->ldap_connection)->isOfType('ldap link');
+
+      /** FIXME: should work...
```
Comment (orthagh) [inline tests/units/AuthLdap.php]: drop it. AuthLdap::prepareInputForUpdate forbid edition of sync_field

File: tests/units/AuthLdap.php [hunk 3]
```diff
@@ -850,4 +865,251 @@ public function testLdapImportUserGroup() {
...
+      $this->boolean($user->getFromDB($user->getID()))->isTrue();
+      $this->array($synchro)
+         ->hasSize(2)
+         ->integer['action']->isIdenticalTo(\AuthLDAP::USER_SYNCHRONIZED)
+         ->variable['id']->isEqualTo($user->getID());
+
+      $this->variable($user->fields['sync_field'])->isEqualTo(42);
+      $this->string($user->fields['name'])->isIdenticalTo('testecuador');
+
+      /** FIXME: should work...
```
Comment (orthagh) [inline tests/units/AuthLdap.php]: drop it. AuthLdap::prepareInputForUpdate forbid edition of sync_field

File: inc/authldap.class.php [hunk 4]
```diff
@@ -1263,6 +1306,58 @@ static function date2ldapTimeStamp($date) {
...
+   * Check if the sync_field is configured for an LDAP server
+   *
+   * @since 9.2
+   * @param integer authldaps_id the LDAP server ID
+   * @return boolean true if configured, false if not configured
+   */
+   public static function isSyncFieldConfigured($authldaps_id) {
+      $authldap = new self();
+      $authldap->getFromDB($authldaps_id);
+      return ($authldap->fields['sync_field'] != null);
```
Comment (orthagh) [inline inc/authldap.class.php]: ``return $authldap->isSyncFieldEnabled();``, no ?

File: inc/authldap.class.php [hunk 5]
```diff
@@ -1617,6 +1769,7 @@ static function getAllUsers(array $options, &$results, &$limitexceeded) {
                                                                $values['end_date']);
             $filter           = "(&$filter $filter_timestamp)";
          }
+
```
Comment (orthagh) [inline inc/authldap.class.php]: drop

File: inc/authldap.class.php [hunk 6]
```diff
@@ -1696,9 +1851,13 @@ static function getAllUsers(array $options, &$results, &$limitexceeded) {
...
             //If user dn exists in DB, it means that user login field has changed
             if (!$tmpuser->getFromDBByDn(toolbox::addslashes_deep($user_infos[$user]["user_dn"]))) {
-               $list[] = ["user"      => $user,
-                               "timestamp" => $user_infos[$user]["timestamp"],
-                               "date_sync" => Dropdown::EMPTY_VALUE];
+               $entry  = ["user"      => $user_infos[$user][$config_ldap->fields['login_field']],
+                          "timestamp" => $user_infos[$user]["timestamp"],
+                          "date_sync" => Dropdown::EMPTY_VALUE];
+               if ($config_ldap->isSyncFieldEnabled()) {
+                  $entry[$config_ldap->getLdapIdentifierToUse()] = $user_infos[$user][$config_ldap->fields['sync_field']];
```
Comment (orthagh) [inline inc/authldap.class.php]: ``$config_ldap->getLdapIdentifierToUse()`` -> ``$field_for_sync``
same with ``$config_ldap->fields['sync_field']``.

So, maybe i'm wrong but the line could be reduced to 
```
 $entry[$field_for_sync] = $user_infos[$user][$field_for_sync];
```

File: inc/authldap.class.php [hunk 7]
```diff
@@ -1600,14 +1745,21 @@ static function getAllUsers(array $options, &$results, &$limitexceeded) {
          $values['order'] = "ASC";
       }
       $ds = $config_ldap->connect();
+      $field_for_sync = $config_ldap->getLdapIdentifierToUse();
```
Comment (orthagh) [inline inc/authldap.class.php]: i count 3 $config_ldap->getDatabaseIdentifierToUse() in the code below.
Maybe like $field_for_sync, we could use a var to improve readability ($field_for_db ?)

File: inc/user.class.php [hunk 8]
```diff
@@ -1390,6 +1400,32 @@ private function getFromLDAPGroupDiscret($ldap_connection, $ldap_method, $userdn
    }
 
 
+   function syncAuthField($ldap_connection, $ldap_method, $userdn, $login) {
```
Comment (trasher) [inline inc/user.class.php]: Does not seems to be used anyways... :)

File: inc/authldap.class.php [hunk 9]
```diff
@@ -2480,24 +2677,33 @@ static function importUserFromServers($options = []) {
...
+            //user login/dn have changed
+            $auth->user->fields['name']      = $login;
+            $auth->user->fields['user_dn']   = $user_dn;
+            $auth->user_present              = true;
          } else { // The user is a new user
             $auth->user_present = $auth->user->getFromDBbyName(addslashes($login));
          }
          $auth->user->getFromLDAP($auth->ldap_connection, $ldap_method, $user_dn, $login,
                                   !$auth->user_present);
+
```
Comment (orthagh) [inline inc/authldap.class.php]: drop

File: inc/authldap.class.php (hunk 7 above)
Comment (trasher) [inline inc/authldap.class.php]: Done :)

**Pull request merged**